Full implementation of rust.workspace()

bonzini opened the pull request:
For the spec see #14639.

This adds a Cargo package object that can provide rustc arguments and dependencies&mdash;but also nice shortcuts `pkg.executable()`, `pkg.library()`, `pkg.proc_macro()`, `pkg.shared_module()` that can basically reduce a Rust crate to just:

```
project('something', 'rust')
cargo = import('rust').workspace()
cargo.package().executable(install: true)
```

[Here](https://gitlab.com/bonzini/libblkio/-/commit/b14230d59d7fcf9323cfad1969307a9418dcb449) is an example of using it for a cdylib composed of three crates.

Based on #15158.
Fixes: #14639

bonzini committed:
modules: rust: add workspace methods returning arguments for build targets

Add rustc_args(), env(), and rust_dependency_map() methods to the
RustPackage class.  They simply delegate to PackageState and
PackageConfiguration.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: rust: implement workspace.package()

Note that, as shown in the testcase, package() works in the subproject
as well.  This means that in the future the Cargo code generator can be
changed to reduce the amount of generated code and instead rely on the
package object.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
test: rust.workspace: add another member to the workspace

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
cargo: use rust.workspace() to build the arguments

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
rust: add to_system_dependency

Move the logic for system dependencies outside Cargo.interpreter and
into the rust module, so that it can be reused by the workspace object.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
cargo: use rust.to_system_dependency

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: allow passing an array for dependency versions

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: rust: invoke subprojects automatically from dependencies()

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: rust: add dependencies() method to package object

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
cargo: populate bin table from src/bin/

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: rust: add package.library/package.proc_macro methods

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: rust: add package.executable

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

jpakkane commented:
> Merge branch 'master' into cargo-object-full

Hmmm?

bonzini commented:
> Hmmm?

It's a draft. :) it will go away when I rebase before making it non-draft. I often have merge commits locally when something builds on top of other PRs.

bonzini committed:
docs: update info on Cargo workspace object

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
modules: rust: add package.override_dependency method

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
dependencies: make arguments to InternalDependency.__init__ optional

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
cargo: use dependencies() method to get dependencies

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

alex-s168 commented:
hmm:
```
rustc -C linker=cc --color=always -C debug-assertions=yes -C overflow-checks=no -C embed-bitcode=no --crate-type proc-macro --cap-lints allow --edition=2021 -g --crate-name thiserror_impl --emit dep-info=subprojects/thiserror-impl-2.0.18/libthiserror_impl+2.so.p/thiserror_impl+2.d --emit link=subprojects/thiserror-impl-2.0.18/libthiserror_impl+2.so -C metadata=08a18b0@@thiserror_impl+2@sha --check-cfg 'cfg(test)' --cfg 'feature="default"' --extern proc_macro --extern proc_macro2=subprojects/proc-macro2-1.0.106/libproc_macro2+1.rlib --extern quote=subprojects/quote-1.0.44/libquote+1.rlib --extern syn=subprojects/syn-2.0.114/libsyn+2.rlib -Lsubprojects/proc-macro2-1.0.106 -Lsubprojects/unicode-ident-1.0.22 -Lsubprojects/quote-1.0.44 -Lsubprojects/syn-2.0.114 -C prefer-dynamic -C link-arg=-Wl,-rpath,{path}/lib/rustlib/x86_64-unknown-linux-gnu/lib ../subprojects/thiserror-impl-2.0.18/src/lib.rs
error: environment variable `CARGO_PKG_VERSION_PATCH` not defined at compile time
  --> ../subprojects/thiserror-impl-2.0.18/src/lib.rs:51:34
   |
51 |             concat!("__private", env!("CARGO_PKG_VERSION_PATCH")),
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Cargo sets build script variables at run time. Use `std::env::var("CARGO_PKG_VERSION_PATCH")` instead
```

(thiserror is a direct dep of a package in my workspace)

bonzini commented:
Add `rust_abi: 'rust'` to both library and override_dependency.

alex-s168 commented:
> override_dependency

Then I get `ERROR: Try to link Rust ABI library 'abc+1' with a non-Rust target 'xyz'`

bonzini commented:
How is XYZ defined?

alex-s168 commented:
> How is XYZ defined?

Well `xyz` is a normal C `static_library`

alex-s168 commented:
anyways I got it working by just having a separate crate for the ffi stuff

alex-s168 commented:
Sure, I'll give it a shot

alex-s168 commented:
yes

bonzini committed:
wrap: create subprojects/ automatically for Cargo subprojects

Since cargo-based resolution derives dependencies from
Cargo.toml/Cargo.lock rather than .wrap files, the directory may not
exist yet.  Do not raise an error if it does not, instead create it
once Cargo.lock and Cargo.toml are loaded.

Suggested-by: Alexander Nutz <alexander.nutz@vxcc.dev>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

alex-s168 commented:
Another annoyance: after running `meson setup --reconfigure`, it always rebuilds all rust libraries...

bonzini committed:
cargo: correct subproject name when git crates are requested

When workspace.subproject() looks up a package, it is currently using
always the NAME-API-rs convention, but that is incorrect for git-method
subprojects (because you don't want the name to change whenever the
API is changed in the git repository).

For external packages, use the subproject name as passed to
_fetch_package_from_subproject and save it for later use in
workspace.subproject().

Fixes: f864690f3 ("cargo: move fetching to _dep_package", 2025-10-14)
Reported-by: Alexander Nutz <alexander.nutz@vxcc.dev>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
cargo: correct subproject name for crates with git source

Use the subproject name from _parse_git_url directly, instead of
treating it as a meson dependency name (which it is not).

Note that while the change affects the "elif dep.git:" branch, that
branch is not visible in the patch because it *keeps calling
_fetch_package_from_subproject* rather than the newly-split
_fetch_package_from_provider.

While at it, move the meson_depname variable out of
_fetch_package, since the caller does not need it.

Fixes: f864690f3 ("cargo: move fetching to _dep_package", 2025-10-14)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini commented:
Ok, the bug is that the Cargo interpreter is using `chumsky` as the subdirectory when cloning, and `chumsky-0.12-rs` when recursing.

Your `chumsky-0.12-rs.wrap` file is then used instead of the git repository, and because it wasn't loaded from Cargo.lock it is configured from scratch - default features and everything.  Remove the `chumsky-0.12-rs.wrap` file and you get the root cause of the bug:

```
ERROR: Cargo subproject chumsky-0.12-rs is buildable: NO

main-lib/meson.build:3:60: ERROR: Neither a subproject directory nor a chumsky-0.12-rs.wrap file was found.
```

This could be separated in a separate PR, for now I'm leaving it here.

bonzini committed:
modules: rust: add package.shared_module method

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
